### PR TITLE
search 7.3 | cross cluster replication | LRDOCS-7402

### PR DESCRIPTION
**Pull request by rbohl**
https://issues.liferay.com/browse/LRDOCS-7402

### CODE REVIEW

**Comment by rbohl** [comment]
@lipusz can you please review my work on the configuration reference article here and let me know if it looks wrong?

https://github.com/rbohl/liferay-learn/pull/62/commits/8c6b2e8f9d3c53d1b3aa1e29142775720d1f14f7

I have this set up locally but not currently with security. If you'd like to see the whole configuration reference in its current state and review it, see

 https://github.com/rbohl/liferay-learn/pull/62/files#diff-073cb7eb9c40feb040d3d14d8b359da2dbd7e379663695db552cf1d78476c8be

Let's start with this, and it will drive the rest of my updates to the articles.

**Comment by rbohl** [comment]
I'll keep working on this but knowing if this is good or needs adjustment will be super helpful so I don't say the wrong thing in a bunch of different places ;)

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/ccr-basic-use-case-config-reference.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+```properties
+productionModeEnabled="true"
+networkHostAddresses=["http://localhost:9200"]
+logExceptionsOnly="false"
+authenticationEnabled="true"
+username="elastic"
+password="liferay"
+httpSSLEnabled="true"
+networkHostAddresses=["https://localhost:9200"]
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/ccr-basic-use-case-config-reference.md:33]: You have 2 `networkHostAddresses` defined: one with and another without `https`. As the rest of the settings are for security, a security is highly recommended, I don't see the point to cover non-secure setups for CCR.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/ccr-basic-use-case-config-reference.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+```
+
+File name: `com.liferay.portal.search.elasticsearch7.configuration.ElasticsearchConnectionConfiguration-remote.config`
+
+File contents:
+
+```properties
+active="true"
+connectionId="remote"
+networkHostAddresses=["http://localhost:9200"]
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/ccr-basic-use-case-config-reference.md:50]: Same: use `https`.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/ccr-basic-use-case-config-reference.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+```
+
+File name: `com.liferay.portal.search.elasticsearch7.configuration.ElasticsearchConnectionConfiguration-ccr.config`
+
+File contents:
+
+```properties
+active="true"
+connectionId="ccr"
+networkHostAddresses=["http://localhost:9202"]
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/ccr-basic-use-case-config-reference.md:67]: Same.

**Comment by lipusz** [comment]
Hey Russ,

I can only continue looking into this after my holiday. I'll return on Jan 6. :christmas_tree:

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+> Elasticsearch 6 Only
+
+For Elasticsearch 6 versions that support Cross-Cluster Replication (6.7+), you must enable [soft deletes](https://www.elastic.co/guide/en/elasticsearch/reference/6.7/ccr-requirements.html) for all existing indexes. This extra hurdle can be avoided if your are on Elasticsearch 7 where soft deletes are enabled by default, so you should [upgrade to Elasticsearch 7](https://help.liferay.com/hc/en-us/articles/360035444872-Upgrading-to-Elasticsearch-7) if at all possible before configuring CCR.
+
+## Enabling Soft Deletes on the System and Company Indexes
+
+Enable soft deletes on the system (`liferay-0`) and company (`liferay-[companyId]`) indexes by adding one line to the _Additional Index Configurations_ setting in Control Panel &rarr; Configuration &rarr; System Settings &rarr; Elasticsearch [Version]:
+ 
+```yaml
+index.soft_deletes.enabled: true
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md:12]: This is enabled by default on ES 7.0

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+```
+
+Alternatively, enable soft deletes in a configuration file named `com.liferay.portal.search.elasticsearch6.configuration.ElasticsearchConfiguration.config` and placed in `[Liferay Home]/osgi/configs`. Give the file these contents:
+
+```properties
+additionalIndexConfigurations="index.soft_deletes.enabled: true"
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md:18]: Not needed.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+```
+
+```note::
+   These steps require performing a full reindex from Control Panel > Configuration > Search, in the Index Actions tab.
+```
+
+## Enabling Soft Deletes on App Indexes
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md:25]: Not needed as we are not going to support CCR on ES 6.x

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+1. Start Liferay DXP (the Leader node).
+ 
+For consistency, this should be done on all DXP cluster nodes residing in your primary (leader) data center.
+ 
+```note::
+    You should review the default settings files each time you `install a new patch <./../../installation-and-upgrades/maintaining-a-liferay-dxp-installation/installing-patches/introduction-to-installing-patches.md>`__ in your DXP environment, and adjust your override files accordingly.
+```
+
+## Existing Deployments: Manually Enabling Soft Deletes
+
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md:86]: Not needed.

Comment (lipusz) [reply docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md:86]: You can remove the all "enable soft deletes" sections completely. In order to preserve this useful material though, you could comment out the content or put it into Grow. Who knows if we will need it at one point in the future. :-)

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+```
+
+### Manually Replicating the Leader Indexes
+
+The replication of indexes in CCR is a _pull_ type operation: it's executed from the local cluster holding the follower indexes. First you'll need to configure the leader/follower relationship between the clusters, then perform the replication on each index.
+
+```tip::
+   Reconfigure the ``kibana.yml`` to connect to your local/follower Elasticsearch cluster:
+
+   ``elasticsearch.hosts: [ "http://localhost:9201" ]``
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md:120]: Use https.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+
+```bash
+com.liferay.portal.search.elasticsearch7.configuration.ElasticsearchConfiguration.config
+```
+
+This file configures the write-enabled connection to the remote Elasticsearch cluster with the leader indexes, including the XPack security settings. Give it these contents:
+
+```properties
+productionModeEnabled="true"
+networkHostAddresses=["http://localhost:9200"]
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md:219]: Use https + configure security.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+```bash
+com.liferay.portal.search.elasticsearch.cross.cluster.replication.internal.configuration.ElasticsearchConnectionConfiguration-follower.config
+```
+
+Give it these contents (security configurations are commented out):
+
+```properties
+connectionId="follower"
+clusterName="LiferayElasticsearchCluster_FOLLOWER"
+networkHostAddress="http://localhost:9201"
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md:234]: There is no `transportAddresses` property in DXP 7.3. I thought this whole documentation is for DXP 7.3 for now. So this config does not work in 7.3. :thinking:

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+transportAddresses=["localhost:9301"]
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md:235]: same

**Comment by rbohl** [comment]
Thanks Tibor! 

I was only looking for a review of the reference for now but appreciate the additional comments.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+#networkHostAddress="https://localhost:9201"
+#authenticationEnabled=B"true"
+#username="elastic"
+#password="liferay"
+#certificateFormat="PEM"
+#sslKeyPath="/PATH/TO/elastic-certificates.key"
+#sslCertificatePath="/PATH/TO/elastic-certificates.crt"
+#sslCertificateAuthoritiesPaths="/PATH/TO/ca.crt"
+#transportSSLVerificationMode="certificate"
+#transportSSLEnabled=B"true"
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-local-follower-data-center.md:246]: same issue, plus, use secure settings by default.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-remote-leader-data-center.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+```bash
+com.liferay.portal.search.elasticsearch7.configuration.ElasticsearchConfiguration.config
+```
+
+Give it these contents:
+
+```properties
+productionModeEnabled="true"
+networkHostAddresses=["http://localhost:9200"]
```
Comment (lipusz) [inline docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-in-a-remote-leader-data-center.md:54]: use https + configure security.

File: docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md (hunk 7 above)
Comment (rbohl) [reply docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication/configuring-ccr-enabling-soft-deletes-on-elasticsearch-6.md:86]: https://github.com/rbohl/liferay-learn/tree/search-7.x-ccr_LRDOCS-7402/docs/dxp/7.x/en/using-search/liferay-enterprise-search/cross-cluster-replication

Yeah, I've forked the branch at this point, so I can just delete sections now, and we'll still have the info preserved in the old branch.

**Pull request closed without merging**